galsim/zernike.py

```diff
@@ -0,0 +1,398 @@
+# Copyright (c) 2012-2017 by the GalSim developers team on GitHub
+# https://github.com/GalSim-developers
+#
+# This file is part of GalSim: The modular galaxy image simulation toolkit.
+# https://github.com/GalSim-developers/GalSim
+#
+# GalSim is free software: redistribution and use in source and binary forms,
+# with or without modification, are permitted provided that the following
+# conditions are met:
+#
+# 1. Redistributions of source code must retain the above copyright notice, this
+#    list of conditions, and the disclaimer given in the accompanying LICENSE
+#    file.
+# 2. Redistributions in binary form must reproduce the above copyright notice,
+#    this list of conditions, and the disclaimer given in the documentation
+#    and/or other materials provided with the distribution.
+#
+"""@file zernike.py
+Module contains code for evaluating and fitting Zernike polynomials
+"""
+
+import numpy as np
+
+from .utilities import LRU_Cache, binomial, horner2d
+
+# Some utilities for working with Zernike polynomials
+# Combinations.  n choose r.
+# See http://stackoverflow.com/questions/3025162/statistics-combinations-in-python
+# This is J. F. Sebastian's answer.
+def _nCr(n, r):
+    if 0 <= r <= n:
+        ntok = 1
+        rtok = 1
+        for t in range(1, min(r, n - r) + 1):
+            ntok *= n
+            rtok *= t
+            n -= 1
+        return ntok // rtok
+    else:
```

Comment:
This branch isn't covered.  Seems like an easy test to add.  Also, maybe consider moving this function to utilities an remove the underscore?

Reply:
Done.  I also did some timing testing on this one.  Turns out that just writing:
```python
def nCr(n, r):
    from math import factorial
    if 0 <= r <= n:
        return factorial(n) // (factorial(r) * factorial(n-r))
    else:
        return 0
```
is usually faster than the more complicated thing that was there before, so I replaced it with the above.  I think the exceptions are when n >> r and n >> 1, in which case the complicated math is faster.  I don't think this should hardly ever occur in GalSim though -- certainly not for Zernike coefficients anyway, which is the only place we currently use this.

Reply:
That feels wrong to me, but I guess it's ok in python, since they have the unlimited size long type.  20! is the largest value that can be stored in an int (C long).  Above that, python switches to its extended precision long.
```
>>> math.factorial(20)
2432902008176640000
>>> math.factorial(21)
51090942171709440000L
>>> type(math.factorial(21))
<type 'long'>
>>> type(math.factorial(20))
<type 'int'>
```
So in C, n>20 would fail with the naive code.  But I would suspect that this is also the transition to where in Python, the naive code gets slower.  Maybe preserve the old code in a branch `if n > 20`?  Or do you think this is not particularly helpful.  (Premature optimization and all that.)

Reply:
This was fun to look into a little further.  Here's a plot showing the log time ratio for the complex (for lack of a better word) algorithm vs the naive algorithm up to n=100.  Reds are where the naive algorithm is faster, blue/greens where the complex algorithm is faster.  

![unknown](https://user-images.githubusercontent.com/3650485/36997605-0808afbc-206f-11e8-8dc4-fdd7372d02fc.png)

There's definitely a transition at n=20, but that's not the dominant effect.  My guess is that the python loop in the complex algorithm (O(min(r, n-r)) is eating a lot of time, while the naive algorithm is O(1) in python once you've memoized the factorials (there are surely some C loops in the arbitrary precision arithmetic though).

So while we could potentially switch algorithms using something like `if n > 20 and min(r, n-r) < n//8`, I don't think it's actually worth doing, given how unlikely this is to be a tall pole anywhere.

Reply:
... looks like the region where the naive algorithm wins does get narrower as n increases.  Here's the same plot up to n=1000, but only doing every 25th n and r:

![unknown-1](https://user-images.githubusercontent.com/3650485/36998602-4c18b406-2072-11e8-9b09-d24dccfce653.png)

Reply:
Succumbing to my premature optimization weakness, I looked into this a bit.  The naive version does seem to get significantly (~30%) slower when n crosses from 20 to 21.  It depends a bit on the value of r whether it becomes slower than the old code or not.

If n >> 20 (like 100 or so), and r is relatively small, then the difference can be quite large in favor of the old code.  I guess that's probably never a use case we'll care about, but a potential concern nonetheless.

Adding an extra if statement does slow things down a bit, especially for the typically small n,r values we use.  But if I tweak the old code a bit to make it more efficient, then I think this is almost always faster than the version with factorial.
```
def nCr(n, r):
    if 0 <= r <= n:
        r = min(r, n-r)
        numer = n
        denom = r
        while r > 1:
            n -= 1
            r -= 1
            numer *= n
            denom *= r
        return numer // denom
    else:
        return 0
```
At least on my laptop, the factorial version was never more than about 10% faster, and this version could be up to a factor of 2 faster (e.g. for r=1 cases) for small n and many times faster for large n (n>30 or so).

Reply:
Weird.  Your posts didn't show up until I hit reload.  Anyway, good to see we both have the same weakness.  :)  Curious what your plot looks like with my modified version.

Reply:
Lol, yep, I'm already on it.  If the fate of the world ever depends on distracting Josh and Mike, now they know how to do it :).

Reply:
I am enjoying this conversation from the outside.  :)

Reply:
Okay, new plots.  Main caveat to following is this is only on my laptop.

Here's the time for each (n, r) up to n=100.

![unknown](https://user-images.githubusercontent.com/3650485/37002891-0540673c-2080-11e8-8483-0d4c3686b9ff.png)

So pretty flat for the naive algorithm, and a pretty significant dependence on n and r for the speedy algorithms.

Same thing as ratios:

![unknown-1](https://user-images.githubusercontent.com/3650485/37002948-30633ed0-2080-11e8-90ab-28e44cd11a6e.png)

So the new speedy algorithm is indeed speedier for min(r, n-r) << n/2, but slower nearer the center.

Zooming out:

![unknown-2](https://user-images.githubusercontent.com/3650485/37002987-4edad9e0-2080-11e8-8693-0ac98266f7a2.png)

Naive depends on n, but not so much on r.  Both speedies scale with both r and n.

![unknown-3](https://user-images.githubusercontent.com/3650485/37003021-711b3cf2-2080-11e8-9330-3f04eaf22212.png)

New and old speedy are about the same in the r, n ~= 100s range.  It's also clear here that, as Mike said, the potential relative speed gains are much larger than the potential speed losses comparing speedy vs naive.

That said, the places where speedy is much speedier are speedy already.  I.e., the fractional improvements are large, but the absolute improvements not so much.  If I just add up all the times for all the (n, r) combinations for each algorithm, I get:

n<100
-------
naive: 10.26
speedy_old: 17.06
speedy_new: 19.16

n < 1000
---------
naive: 38.84
speedy_old: 36.99
speedy_new: 37.01

So assuming a uniform prior over (n, r) with n<1000 (reasonable?), it really doesn't make much difference to the expected value of calculation time.  Over n<100, then naive clearly wins.

Reply:
Very interesting.  I wonder if math.factorial got faster in python 3.  My results were using python 2.7, and your plots are red in places where the new code definitely won on my laptop.  

E.g. nCr(10,9) on my laptop is more than a factor of 2 faster with the new algorithm, but it shows up as pink in your t_speedy_new/t_naive plot.  And there were only a few combinations I could find with n < 20 where the naive algorithm was faster.  (Mostly where n ~ 2r as you would expect.)

Anyway, fine to leave it with the factorial version.  Python 3 is the future anyway, so might as well optimize for that to the extent we should bother optimizing this at all.

Reply:
>I wonder if math.factorial got faster in python 3.

Yep, this looks like it is the case.  Here's the first plot again, but in py2.7

![unknown-4](https://user-images.githubusercontent.com/3650485/37007399-fc992666-2091-11e8-9e08-508104824894.png)

So a strong dependence on n instead of being mostly flat (for the naive algorithm).

>Python 3 is the future anyway, so might as well optimize for that to the extent we should bother optimizing this at all.

Sounds good to me.